src/charm.py

```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
-        pgbouncer.ini files in a more sophisticated way.
+    def _read_pgb_config(self) -> pgb.PgbConfig:
+        """Get config object from pgbouncer.ini file stored on container.
 
-        Args:
-            users: a dictionary of usernames and passwords
         Returns:
-            A multiline string defining a valid pgbouncer.ini file
+            PgbConfig object containing pgbouncer config.
         """
-        if users is None:
-            users = self._get_userlist_from_container()
-
-        return f"""[databases]
-{self.config["pgb_databases"]}
-
-[pgbouncer]
-listen_port = {self.config["pgb_listen_port"]}
-listen_addr = {self.config["pgb_listen_address"]}
-auth_type = md5
-auth_file = userlist.txt
-logfile = pgbouncer.log
-pidfile = pgbouncer.pid
-admin_users = {",".join(users.keys())}"""
-
-    def _push_userlist(self, users: Dict = None, reload_pgbouncer: bool = False) -> None:
+        try:
+            config = self._read_file(INI_PATH)
+            return pgb.PgbConfig(config)
+        except FileNotFoundError:
+            logger.error("pgbouncer config not found")
+
+    def _render_userlist(self, userlist: Dict, reload_pgbouncer: bool = False) -> None:
         """Generate userlist.txt from the given userlist dict, and deploy it to the container.
 
         Args:
-            users: a dictionary of usernames and passwords
+            userlist: a dictionary of usernames and passwords
             reload_pgbouncer: A boolean defining whether or not to reload the pgbouncer application
                 in the container. When config files are updated, pgbouncer must be restarted for
                 the changes to take effect. However, these config updates can be done in batches,
                 minimising the amount of necessary restarts.
         """
-        pgb_container = self.unit.get_container(self._pgbouncer_container)
-        if users is None:
-            users = self._get_userlist_from_container()
-        userlist = self._generate_userlist(users)
-
-        try:
-            # Check that we're not updating this file unnecessarily
-            if pgb_container.pull(USERLIST_PATH).read() == userlist:
-                logger.info("updated userlist does not modify existing pgbouncer userlist")
-                return
-        except FileNotFoundError:
-            # there is no existing userlist.txt file, so carry on and add one.
-            pass
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "Unable to push config to container - container unavailable."
+            )
+            return
 
         pgb_container.push(
             USERLIST_PATH,
-            userlist,
-            user=self._pgbouncer_user,
-            permissions=0o600,
+            pgb.generate_userlist(userlist),
+            user=PG_USER,
+            permissions=0o777,
```

Comment:
Is `777` really needed? PgBouncer by default creates this file with `600` when installed through apt.

Reply:
You're right, this has been reverted back to 600